src/AstGen.zig

```diff
@@ -6029,16 +6029,15 @@ fn switchExpr(
     const cond_ty_inst = try parent_gz.addUnNode(.typeof, cond, operand_node);
     const item_rl: ResultLoc = .{ .ty = cond_ty_inst };
 
-    // These contain the data that goes into the `extra` array for the SwitchBlock/SwitchBlockMulti.
-    // This is the optional else prong body.
-    var special_case_payload = ArrayListUnmanaged(u32){};
-    defer special_case_payload.deinit(gpa);
-    // This is all the scalar cases.
-    var scalar_cases_payload = ArrayListUnmanaged(u32){};
-    defer scalar_cases_payload.deinit(gpa);
-    // Same deal, but this is only the `extra` data for the multi cases.
-    var multi_cases_payload = ArrayListUnmanaged(u32){};
-    defer multi_cases_payload.deinit(gpa);
+    // This contains the data that goes into the `extra` array for the SwitchBlock/SwitchBlockMulti,
+    // except the first cases_nodes.len slots are a table that indexes payloads later in the array, with
+    // the special case index coming first, then scalar_case_len indexes, then multi_cases_len indexes
+    var payloads = ArrayListUnmanaged(u32){};
+    defer payloads.deinit(gpa);
+    const scalar_case_table: u32 = @boolToInt(special_prong != .none);
+    const multi_case_table = scalar_case_table + scalar_cases_len;
+    const case_table_len = multi_case_table + multi_cases_len;
+    try payloads.resize(gpa, case_table_len);
```

Comment:
This will usually over-allocate (see https://github.com/ziglang/zig/pull/10076 for more details). It would be better to call `initCapacity` instead of `resize` so that `resize` doesn't over-allocate (because `resize` will calculate a `better_capacity` larger than the requested capacity, while `initCapacity` won't).

```suggestion
    const scalar_case_table: u32 = @boolToInt(special_prong != .none);
    const multi_case_table = scalar_case_table + scalar_cases_len;
    const case_table_len = multi_case_table + multi_cases_len;
    var payloads = ArrayListUnmanaged(u32).initCapacity(gpa, case_table_len);
    defer payloads.deinit(gpa);
    payloads.items.len = case_table_len;
```